Clarify impure in systems of equations. (#3571)
modelica/ModelicaSpecification: chapters/functions.tex

  If a function is declared as \lstinline!impure! any function extending from it shall be declared as \lstinline!impure!.
\item
  A deprecated semantics is that external functions (and functions defined in Modelica directly or indirectly calling them) without \lstinline!pure! or \lstinline!impure! keyword are assumed to be impure, but without any restriction on calling them.

>   Similarly they can be part of systems of equations without restrictions.

Reviewer: If we manage to write the text clearly, there shouldn't be a need for this sentence, as the systems-of-equations rule should be perceived as one of the calling restrictions.

The reason is that solving systems of equations generally requires expressions to be evaluated an unknown number of times.
This includes the special handling of \lstinline!when initial()! during initialization.

There are two ways in which an impure function could be called in a system of equations, namely in the deprecated case of external functions assumed to be impure, and when using \lstinline!pure($\ldots$)! to call an impure function from within a pure function.
The side-effect semantics are undefined in this case.

> Specifically the number of calls with external side-effects is unspecified.

Reviewer: Missing comma:
```suggestion
Specifically, the number of calls with external side-effects is unspecified.
```

The reason is that solving systems of equations generally requires expressions to be evaluated an unknown number of times.
This includes the special handling of \lstinline!when initial()! during initialization.

> There are two ways in which an impure function could be called in a system of equations,
> namely in the deprecated case of external functions assumed to be impure, and when using \lstinline!pure($\ldots$)! to call an impure function from within a pure function.

Reviewer: Sentence-based line breaking:
```suggestion
There are two ways in which an impure function could be called in a system of equations, namely in the deprecated case of external functions assumed to be impure, and when using \lstinline!pure($\ldots$)! to call an impure function from within a pure function.
```

> It is still possible to call an impure function in a system of equations in specific ways.
> Specifically for the deprecated case of external functions assumed to be impure, and when using \lstinline!pure(expression)! to call an impure function in a pure function.

Reviewer: Looks good technically, but could be written in a slightly more readable fashion.  For example, it would be nice to reduce the number of _specific…_ in the section.
```suggestion
There are two ways in which an impure function could be called in a system of equations,
namely in the deprecated case of external functions assumed to be impure, and when using \lstinline!pure($\ldots$)! to call an impure function from within a pure function.
```
(The alternatve to `\ldots` above would be `\mathit{expression}`.)

> The side-effect semantics are undefined in this case.

Reviewer: Just asking: Is it clear that _this case_ refers to both ways in the impure function could be called?

Reviewer: OK, let me turn that into a concrete suggestion:
```suggestion
The side-effect semantics are undefined in these cases.
```
or avoid _this case_ by writing:
```suggestion
The side-effect semantics of the function call are then undefined.
```

The reason is that solving systems of equations generally requires expressions to be evaluated an unknown number of times.
This includes the special handling of \lstinline!when initial()! during initialization.

> \begin{nonnormative}

Author: ```suggestion
```

It is still possible to call an impure function in a system of equations in specific ways.
Specifically for the deprecated case of external functions assumed to be impure, and when using \lstinline!pure(expression)! to call an impure function in a pure function.
The side-effect semantics are undefined in this case.
Specifically the number of calls with external side-effects is unspecified.
However, for impure functions where the outputs only depend on the inputs the system of equations should be solved correctly.

> \end{nonnormative}

Author: ```suggestion
```

A function having a formal function parameter that is \lstinline!impure! must be marked \lstinline!pure! or \lstinline!impure!.

\begin{nonnormative}

> Comment: The semantics are undefined if the function call of an impure function is part of an algebraic loop.

Reviewer: This is the only _Comment:_ at the start of a non-normative section.  Since all non-normative content is some sort of content, we should remove it:
```suggestion
The semantics are undefined if the function call of an impure function is part of an algebraic loop.
```

> Comment: The semantics are undefined if the function call of an impure function is part of an algebraic loop.
> This can occur for the deprecated case of external functions assumed to be pure, or when using \lstinline!pure(expression)! to call an impure function in a pure function.

Author: ```suggestion
Comment: The semantics are undefined if the function call of an impure function is part of an algebraic loop.
This can occur for the deprecated case of external functions assumed to be impure, or when using \lstinline!pure(expression)! to call an impure function in a pure function.
```
Typo!

> Comment: The semantics are undefined if the function call of an impure function is part of an algebraic loop.
> This can occur for the deprecated case of external functions assumed to be impure, or when using \lstinline!pure(expression)! to call an impure function in a pure function.

Author: ```suggestion
Comment: The exact semantics are undefined if the function call of an impure function is part of an algebraic loop.
Specifically the number of calls with external side-effects is unspecified.
However, for impure functions where the outputs only depend on the inputs the algebraic loop should be solved correctly.
This can occur for the deprecated case of external functions assumed to be impure, or when using \lstinline!pure(expression)! to call an impure function in a pure function.
```
I believe this was the intended idea.

Not a general undefined-behavior in C-terms (with nasal demons etc), but just the pragmatic point that you cannot solve an algebraic loop with an impure function that e.g., returns the current time in nano-seconds, and that you have no control of the number of external side-effects generated.

That's why it is non-normative.

Reviewer: How about speaking of _side-effect semantics_?  Wouldn't that then be nicely elaborated by the outputs-only-depending-on-inputs rule?

Author: Ok, makes sense.

Author: Have now done that.
